src/core/service.c

```diff
@@ -2227,6 +2305,103 @@ static int service_serialize(Unit *u, FILE *f, FDSet *fds) {
         return 0;
 }
 
+static int service_deserialize_exec_command(Unit *u, const char *key, const char *value) {
+        Service *s = SERVICE(u);
+        int r;
+        unsigned index, i;
+        bool control, found = false;
+        ServiceExecCommand id;
+        ExecCommand *command = NULL;
+        _cleanup_free_ char *args = NULL, *path = NULL;
+        _cleanup_strv_free_ char **argv = NULL;
+
+        enum ExecCommandState {
+                STATE_EXEC_COMMAND_TYPE,
+                STATE_EXEC_COMMAND_ID,
+                STATE_EXEC_COMMAND_PATH,
+                STATE_EXEC_COMMAND_ARGS,
+                _STATE_EXEC_COMMAND_MAX,
+                _STATE_EXEC_COMMAND_INVALID = -1,
+        } state;
+
+        assert(s);
+        assert(key);
+        assert(value);
+
+        control = streq(key, "control-command");
+
+        state = STATE_EXEC_COMMAND_TYPE;
+
+        for (;;) {
+                _cleanup_free_ char *arg = NULL;
+
+                r = extract_first_word(&value, &arg, NULL, EXTRACT_CUNESCAPE);
+                if (r == 0)
+                        break;
+                else if (r < 0)
+                        return r;
+
+                switch (state) {
+                case STATE_EXEC_COMMAND_TYPE:
+                        id = service_exec_command_from_string(arg);
+                        if (id < 0)
+                                return -EINVAL;
+
+                        state = STATE_EXEC_COMMAND_ID;
+                        break;
+                case STATE_EXEC_COMMAND_ID:
+                        r = safe_atou(arg, &index);
+                        if (r < 0)
+                                return -EINVAL;
+
+                        state = STATE_EXEC_COMMAND_PATH;
+                        break;
+                case STATE_EXEC_COMMAND_PATH:
+                        path = arg;
+                        arg = NULL;
+                        state = STATE_EXEC_COMMAND_ARGS;
+
+                        if (!path_is_absolute(path))
+                                return -EINVAL;
+                        break;
+                case STATE_EXEC_COMMAND_ARGS:
+                        r = strv_extend(&argv, arg);
+                        if (r < 0)
+                                return -ENOMEM;
+                        break;
+                default:
+                        assert_not_reached("Unknown error at deserialization of exec command");
+                        break;
+                }
+        }
+
+        /* Let's check whether exec command on given offset matches data that we just deserialized */
+        for (command = s->exec_command[id], i = 0; command; command = command->command_next, i++) {
+                if (i != index)
```

Comment:
Not that it matters much in practice, but `index` could be uninitialized here. Also `path` could be NULL.

In fact, using a state machine for this seems overkill, since there's only a single trivial pass possible. Why not do something like this:
```C
_cleanup_free_ char *idstr = NULL, *indexstr = NULL, *path = NULL;
char *arg;
r = extract_first_word(&value, &idstr, NULL, EXTRACT_CUNESCAPE);
if (r <= 0)
     return r < 0 ? r : -EUCLEAN;
id = service_exec_command_from_string(idstr);
if (id < 0)
       return -EINVAL;
r = extract_first_word(&value, &indexstr, NULL, EXTRACT_CUNESCAPE);
if (r <= 0)
     return r < 0 ? r : -EUCLEAN;
 r = safe_atou(indexstr, &index);
if (r < 0)
            return -EINVAL;
r = extract_first_word(&value, &path, NULL, EXTRACT_CUNESCAPE);
if (r <= 0)
     return r < 0 ? r : -EUCLEAN;
if (!path_is_absolute(path))
     return -EINVAL;
while ((r = extract_first_word(&value, &arg, NULL, EXTRACT_CUNESCAPE) > 0) {
     r = strv_push(&argv, arg);
     if (r < 0)
            break;
}
if (r < 0)
     return r;
...
```
That's both shorter and imho more bullet-proof.